*** /tmp/MAC_tx_Ctrl.v	2006-06-25 06:59:00.000000000 +0200
--- MAC_tx_Ctrl.v	2011-10-09 11:43:32.544538167 +0200
***************
*** 174,179 ****
--- 174,180 ----
  parameter       StateSwitchNext     =4'd12;
  parameter       StateDefer          =4'd13;
  parameter       StateSendPauseFrame =4'd14;
+ parameter       StateIdle2          =4'd15;
  
  reg [3:0]       Current_state   /*synthesis syn_keep=1 */;
  reg [3:0]       Next_state;
***************
*** 277,282 ****
--- 278,285 ----
                  else
                      Next_state=Current_state;           
              StateIdle:
+                 Next_state=StateIdle2;
+             StateIdle2:
                  if (!FullDuplex&&CRS)
                      Next_state=StateDefer;
                  else if (pause_apply)
