diff -rup sdcc-src-3.1.0/Makefile.common.in sdcc-src-3.1.0-pblaze/Makefile.common.in
--- sdcc-src-3.1.0/Makefile.common.in	2011-10-20 11:10:56.000000000 +0200
+++ sdcc-src-3.1.0-pblaze/Makefile.common.in	2011-12-04 13:58:18.197756400 +0100
@@ -51,6 +51,7 @@ OPT_DISABLE_Z80   = @OPT_DISABLE_Z80@
 OPT_DISABLE_Z180  = @OPT_DISABLE_Z180@
 OPT_DISABLE_R2K   = @OPT_DISABLE_R2K@
 OPT_DISABLE_GBZ80  = @OPT_DISABLE_GBZ80@
+OPT_DISABLE_PBLAZE = @OPT_DISABLE_PBLAZE@
 
 OPT_DISABLE_UCSIM     = @OPT_DISABLE_UCSIM@
 OPT_DISABLE_DEVICE_LIB= @OPT_DISABLE_DEVICE_LIB@
diff -rup sdcc-src-3.1.0/configure sdcc-src-3.1.0-pblaze/configure
--- sdcc-src-3.1.0/configure	2011-10-25 20:43:54.000000000 +0200
+++ sdcc-src-3.1.0-pblaze/configure	2011-12-05 22:37:52.945707300 +0100
@@ -620,6 +620,7 @@ OPT_DISABLE_R2K
 OPT_DISABLE_Z180
 OPT_DISABLE_Z80
 OPT_DISABLE_MCS51
+OPT_DISABLE_PBLAZE
 non_free_lib_dir_suffix
 lib_dir_suffix
 non_free_include_dir_suffix
@@ -716,6 +717,7 @@ enable_pic16_port
 enable_hc08_port
 enable_avr_port
 enable_xa51_port
+enable_pblaze_port
 enable_ucsim
 enable_device_lib
 enable_packihx
@@ -1373,6 +1375,7 @@ Optional Features:
   --disable-pic14-port    Excludes the PIC14 port
   --disable-pic16-port    Excludes the PIC16 port
   --disable-hc08-port     Excludes the HC08 port
+  --disable-pblaze-port   Excludes the PBLAZE port
   --enable-avr-port       Includes the AVR port (disabled by default)
   --enable-xa51-port      Includes the XA51 port (disabled by default)
   --disable-ucsim         Disables configuring and building of ucsim
@@ -6580,6 +6583,32 @@ _ACEOF
 
 
 
+  # Check whether --enable-pblaze-port was given.
+if test "${enable_pblaze_port+set}" = set; then :
+  enableval=$enable_pblaze_port;
+fi
+
+
+  if test "$enable_pblaze_port" = "no"; then
+    OPT_DISABLE_PBLAZE=1
+  else
+    enable_pblaze_port="yes"
+    OPT_DISABLE_PBLAZE=0
+  fi
+
+
+cat >>confdefs.h <<_ACEOF
+#define OPT_DISABLE_PBLAZE $OPT_DISABLE_PBLAZE
+_ACEOF
+
+
+
+  echo pblaze >>ports.all
+  if test $OPT_DISABLE_PBLAZE = 0; then
+    echo pblaze >>ports.build
+  fi
+
+
   # Check whether --enable-ucsim was given.
 if test "${enable_ucsim+set}" = set; then :
   enableval=$enable_ucsim;
@@ -7068,6 +7097,10 @@ if test $OPT_DISABLE_Z80 = 0 || test $OP
 
 fi
 
+if test $OPT_DISABLE_PBLAZE = 0; then
+  ac_config_files="$ac_config_files src/pblaze/Makefile"
+fi
+
 if test $OPT_DISABLE_Z80 = 0 || test $OPT_DISABLE_Z180 = 0; then
   ac_config_files="$ac_config_files sdas/asz80/Makefile"
 
@@ -7797,6 +7830,7 @@ do
     "src/hc08/Makefile") CONFIG_FILES="$CONFIG_FILES src/hc08/Makefile" ;;
     "sdas/as6808/Makefile") CONFIG_FILES="$CONFIG_FILES sdas/as6808/Makefile" ;;
     "device/lib/hc08/Makefile") CONFIG_FILES="$CONFIG_FILES device/lib/hc08/Makefile" ;;
+    "src/pblaze/Makefile") CONFIG_FILES="$CONFIG_FILES src/pblaze/Makefile" ;;
     "src/mcs51/Makefile") CONFIG_FILES="$CONFIG_FILES src/mcs51/Makefile" ;;
     "sdas/as8051/Makefile") CONFIG_FILES="$CONFIG_FILES sdas/as8051/Makefile" ;;
     "device/lib/mcs51/Makefile") CONFIG_FILES="$CONFIG_FILES device/lib/mcs51/Makefile" ;;
@@ -8805,6 +8839,7 @@ sdcc ${VERSION} is now configured for
     z80                 ${enable_z80_port}
     z180                ${enable_z180_port}
     r2k                 ${enable_r2k_port}
+    pblaze              ${enable_pblaze_port}
 
   Disable packihx:      ${OPT_DISABLE_PACKIHX}
   Disable ucsim:        ${OPT_DISABLE_UCSIM}
@@ -8865,6 +8900,7 @@ sdcc ${VERSION} is now configured for
     z80                 ${enable_z80_port}
     z180                ${enable_z180_port}
     r2k                 ${enable_r2k_port}
+    pblaze              ${enable_pblaze_port}
 
   Disable packihx:      ${OPT_DISABLE_PACKIHX}
   Disable ucsim:        ${OPT_DISABLE_UCSIM}
diff -rup sdcc-src-3.1.0/configure.in sdcc-src-3.1.0-pblaze/configure.in
--- sdcc-src-3.1.0/configure.in	2011-10-25 20:43:54.000000000 +0200
+++ sdcc-src-3.1.0-pblaze/configure.in	2011-12-04 14:42:23.262045400 +0100
@@ -815,6 +815,7 @@ AC_DO_PORT(ds400, ds400, DS400, [Exclude
 AC_DO_PORT(pic14, pic14, PIC14, [Excludes the PIC14 port])
 AC_DO_PORT(pic16, pic16, PIC16, [Excludes the PIC16 port])
 AC_DO_PORT(hc08,  hc08,  HC08,  [Excludes the HC08 port])
+AC_DO_PORT(pblaze,  pblaze,  PBLAZE,  [Excludes the PBLAZE port])
 
 # Unsupported targets
 AC_DO_PORT_ENABLER(avr,   avr,   AVR,   [Includes the AVR port (disabled by default)])
@@ -906,6 +907,10 @@ if test $OPT_DISABLE_Z80 = 0 || test $OP
   AC_CONFIG_FILES([src/z80/Makefile])
 fi
 
+if test $OPT_DISABLE_PBLAZE = 0; then
+  AC_CONFIG_FILES([src/pblaze/Makefile])
+fi
+
 if test $OPT_DISABLE_Z80 = 0 || test $OPT_DISABLE_Z180 = 0; then
   AC_CONFIG_FILES([sdas/asz80/Makefile])
   test $OPT_DISABLE_DEVICE_LIB = 0 && AC_CONFIG_FILES([device/lib/z80/Makefile
@@ -991,6 +996,7 @@ sdcc ${VERSION} is now configured for
     z80                 ${enable_z80_port}
     z180                ${enable_z180_port}
     r2k                 ${enable_r2k_port}
+    pblaze              ${enable_pblaze_port}
 
   Disable packihx:      ${OPT_DISABLE_PACKIHX}
   Disable ucsim:        ${OPT_DISABLE_UCSIM}
Only in sdcc-src-3.1.0-pblaze/device/examples: pblaze
diff -rup sdcc-src-3.1.0/device/include/Makefile.in sdcc-src-3.1.0-pblaze/device/include/Makefile.in
--- sdcc-src-3.1.0/device/include/Makefile.in	2011-10-12 16:09:15.000000000 +0200
+++ sdcc-src-3.1.0-pblaze/device/include/Makefile.in	2011-12-05 22:46:57.573665600 +0100
@@ -71,6 +71,10 @@ install: all installdirs
 	    done; \
 	  done; \
 	fi
+	# picoblaze 
+	if [ "`grep pblaze $(top_builddir)/ports.build`" = pblaze ]; then \
+	  $(CP) $(srcdir)/pblaze/*.h $(sdcc_includedir)/pblaze; \
+	fi
 	find $(sdcc_includedir) -type d -name '.svn' -exec rm -rf {} \;
 	# correct file modes
 	find $(sdcc_includedir) -type f -exec chmod 644 {} \;
@@ -96,7 +100,7 @@ installcheck:
 # ---------------------------------
 installdirs:
 	mkdir -p $(sdcc_includedir)
-	for target in mcs51 ds390 ds400 pic14 pic16 z80 z180 gbz80 hc08; \
+	for target in mcs51 ds390 ds400 pic14 pic16 z80 z180 gbz80 hc08 pblaze; \
 	do \
 	  if [ -d $(srcdir)/$${target} ]; \
 	  then \
Only in sdcc-src-3.1.0-pblaze/device/include: pblaze
diff -rup sdcc-src-3.1.0/device/lib/Makefile.in sdcc-src-3.1.0-pblaze/device/lib/Makefile.in
--- sdcc-src-3.1.0/device/lib/Makefile.in	2011-11-01 17:35:46.000000000 +0100
+++ sdcc-src-3.1.0-pblaze/device/lib/Makefile.in	2011-12-04 13:55:53.223464300 +0100
@@ -70,6 +70,7 @@ OPT_DISABLE_XA51  = @OPT_DISABLE_XA51@
 OPT_DISABLE_Z80   = @OPT_DISABLE_Z80@
 OPT_DISABLE_Z180  = @OPT_DISABLE_Z180@
 OPT_DISABLE_R2K   = @OPT_DISABLE_R2K@
+OPT_DISABLE_PBLAZE= @OPT_DISABLE_PBLAZE@
 
 SOURCES_FLOAT = $(COMMON_FLOAT) \
   _fscmp.c \
diff -rup sdcc-src-3.1.0/device/non-free/lib/Makefile.in sdcc-src-3.1.0-pblaze/device/non-free/lib/Makefile.in
--- sdcc-src-3.1.0/device/non-free/lib/Makefile.in	2011-10-25 20:43:54.000000000 +0200
+++ sdcc-src-3.1.0-pblaze/device/non-free/lib/Makefile.in	2011-12-04 13:57:07.901735700 +0100
@@ -68,6 +68,7 @@ OPT_DISABLE_PIC14 = @OPT_DISABLE_PIC14@
 OPT_DISABLE_PIC16 = @OPT_DISABLE_PIC16@
 OPT_DISABLE_XA51  = @OPT_DISABLE_XA51@
 OPT_DISABLE_Z80   = @OPT_DISABLE_Z80@
+OPT_DISABLE_PBLAZE= @OPT_DISABLE_PBLAZE@
 
 SOURCES_FLOAT = $(COMMON_FLOAT) \
   _fscmp.c \
Only in sdcc-src-3.1.0-pblaze/doc: sdcc_pblaze_instructions.txt
diff -rup sdcc-src-3.1.0/sdcc_vc_in.h sdcc-src-3.1.0-pblaze/sdcc_vc_in.h
--- sdcc-src-3.1.0/sdcc_vc_in.h	2011-06-10 23:19:02.000000000 +0200
+++ sdcc-src-3.1.0-pblaze/sdcc_vc_in.h	2011-12-04 15:18:59.087639400 +0100
@@ -77,6 +77,7 @@
 #undef OPT_DISABLE_PIC14
 #undef OPT_DISABLE_PIC16
 #define OPT_DISABLE_XA51	1
+#undef OPT_DISABLE_PBLAZE
 
 #endif  /* SDCC_VC_HEADER */
 
diff -rup sdcc-src-3.1.0/sdccconf_in.h sdcc-src-3.1.0-pblaze/sdccconf_in.h
--- sdcc-src-3.1.0/sdccconf_in.h	2011-10-09 20:21:10.000000000 +0200
+++ sdcc-src-3.1.0-pblaze/sdccconf_in.h	2011-12-04 13:59:25.373598600 +0100
@@ -138,6 +138,9 @@
 #undef OPT_DISABLE_R2K
 
 /* XXX */
+#undef OPT_DISABLE_PBLAZE
+
+/* XXX */
 #undef OPT_DISABLE_SDCDB
 
 /* XXX */
diff -rup sdcc-src-3.1.0/src/SDCCmain.c sdcc-src-3.1.0-pblaze/src/SDCCmain.c
--- sdcc-src-3.1.0/src/SDCCmain.c	2011-10-09 20:21:10.000000000 +0200
+++ sdcc-src-3.1.0-pblaze/src/SDCCmain.c	2011-12-04 14:01:09.673564200 +0100
@@ -340,6 +340,9 @@ static PORT *_ports[] = {
 #if !OPT_DISABLE_HC08
   &hc08_port,
 #endif
+#if !OPT_DISABLE_PBLAZE
+  &pblaze_port,
+#endif
 };
 
 #define NUM_PORTS (sizeof(_ports)/sizeof(_ports[0]))
Only in sdcc-src-3.1.0-pblaze/src: json
Only in sdcc-src-3.1.0-pblaze/src: pblaze
diff -rup sdcc-src-3.1.0/src/port.h sdcc-src-3.1.0-pblaze/src/port.h
--- sdcc-src-3.1.0/src/port.h	2011-10-09 20:21:10.000000000 +0200
+++ sdcc-src-3.1.0-pblaze/src/port.h	2011-12-04 15:56:51.809631700 +0100
@@ -22,6 +22,7 @@
 #define TARGET_ID_HC08     11
 #define TARGET_ID_Z180     12
 #define TARGET_ID_R2K      13
+#define TARGET_ID_PBLAZE   14
 
 /* Macro to test the target we are compiling for.
    Can only be used after SDCCmain has defined the port
@@ -38,6 +39,7 @@
 #define TARGET_IS_XA51     (port->id == TARGET_ID_XA51)
 #define TARGET_IS_HC08     (port->id == TARGET_ID_HC08)
 #define TARGET_IS_R2K      (port->id == TARGET_ID_R2K)
+#define TARGET_IS_PBLAZE   (port->id == TARGET_ID_PBLAZE)
 
 #define TARGET_MCS51_LIKE  (TARGET_IS_MCS51 || TARGET_IS_DS390 || TARGET_IS_DS400)
 #define TARGET_Z80_LIKE    (TARGET_IS_Z80 || TARGET_IS_Z180 || TARGET_IS_GBZ80 || TARGET_IS_R2K)
@@ -427,5 +429,8 @@ extern PORT ds400_port;
 #if !OPT_DISABLE_HC08
 extern PORT hc08_port;
 #endif
+#if !OPT_DISABLE_PBLAZE
+extern PORT pblaze_port;
+#endif
 
 #endif /* PORT_INCLUDE */
