diff --git a/arch/powerpc/include/asm/reg.h b/arch/powerpc/include/asm/reg.h
index c6d1ab6..e2caf0d 100644
--- a/arch/powerpc/include/asm/reg.h
+++ b/arch/powerpc/include/asm/reg.h
@@ -111,7 +111,8 @@
 #define MSR_KERNEL	(MSR_ME|MSR_RI|MSR_IR|MSR_DR)
 #endif
 
-#define MSR_USER	(MSR_KERNEL|MSR_PR|MSR_EE)
+#define MSR_USER	(MSR_KERNEL|MSR_PR|MSR_EE|0x02080000)
+//#define MSR_USER	(MSR_KERNEL|MSR_PR|MSR_EE|0x02080000|(1<<6)|(1<<18))
 #endif
 
 /* Floating Point Status and Control Register (FPSCR) Fields */
