diff -ruN vgabios-0.7a/Makefile vgabios-0.7a-ao486/Makefile
--- vgabios-0.7a/Makefile	2011-07-19 18:59:50.000000000 +0200
+++ vgabios-0.7a-ao486/Makefile	2014-03-30 14:39:25.000000000 +0200
@@ -31,7 +31,8 @@
 VBE_FILES := vbe.h vbe.c vbetables.h
 
 # build flags
-vgabios.bin              : VGAFLAGS := -DVBE -DPCIBIOS
+vgabios.bin              : VGAFLAGS :=
+#-DVBE -DPCIBIOS
 vgabios.debug.bin        : VGAFLAGS := -DVBE -DPCIBIOS -DDEBUG
 vgabios-cirrus.bin       : VGAFLAGS := -DCIRRUS -DPCIBIOS
 vgabios-cirrus.debug.bin : VGAFLAGS := -DCIRRUS -DPCIBIOS -DCIRRUS_DEBUG
@@ -43,7 +44,8 @@
 vgabios-cirrus.debug.bin : DISTNAME := VGABIOS-lgpl-latest.cirrus.debug.bin
 
 # dependencies
-vgabios.bin              : $(VGA_FILES) $(VBE_FILES) biossums
+vgabios.bin              : $(VGA_FILES) biossums
+#$(VGA_FILES) $(VBE_FILES) biossums
 vgabios.debug.bin        : $(VGA_FILES) $(VBE_FILES) biossums
 vgabios-cirrus.bin       : $(VGA_FILES) clext.c biossums
 vgabios-cirrus.debug.bin : $(VGA_FILES) clext.c biossums
